internal/supervisor/group.go

```diff
@@ -0,0 +1,113 @@
+package supervisor
+
+import (
+	log "github.com/sirupsen/logrus"
+
+	"github.com/mattermost/mattermost-cloud/internal/store"
+	"github.com/mattermost/mattermost-cloud/model"
+)
+
+// groupStore abstracts the database operations required to query groups.
+type groupStore interface {
+	GetUnlockedGroupsPendingWork() ([]*model.Group, error)
+	GetGroupRollingMetadata(groupID string) (*store.GroupRollingMetadata, error)
+	LockGroup(groupID, lockerID string) (bool, error)
+	UnlockGroup(groupID, lockerID string, force bool) (bool, error)
+
+	GetInstallation(installationID string, includeGroupConfig, includeGroupConfigOverrides bool) (*model.Installation, error)
+	UpdateInstallationState(id, state string) error
+	LockInstallation(installationID, lockerID string) (bool, error)
+	UnlockInstallation(installationID, lockerID string, force bool) (bool, error)
+}
+
+// GroupSupervisor finds installations belonging to groups that need to have
+// their configuration reconciled to match a new group configuration setting.
+//
+// The degree of parallelism is controlled by a weighted semaphore, intended to
+// be shared with other clients needing to coordinate background jobs.
+type GroupSupervisor struct {
+	store      groupStore
+	instanceID string
+	logger     log.FieldLogger
+}
+
+// NewGroupSupervisor creates a new GroupSupervisor.
+func NewGroupSupervisor(store groupStore, instanceID string, logger log.FieldLogger) *GroupSupervisor {
+	return &GroupSupervisor{
+		store:      store,
+		instanceID: instanceID,
+		logger:     logger,
+	}
+}
+
+// Do looks for work to be done on any pending groups and attempts to schedule
+// the required work.
+func (s *GroupSupervisor) Do() error {
+	groups, err := s.store.GetUnlockedGroupsPendingWork()
+	if err != nil {
+		s.logger.WithError(err).Warn("Failed to query for groups")
+		return nil
+	}
+
+	for _, group := range groups {
+		s.Supervise(group)
+	}
+
+	return nil
+}
+
+// Supervise schedules the required work on the given group.
+func (s *GroupSupervisor) Supervise(group *model.Group) {
+	logger := s.logger.WithFields(log.Fields{
+		"group": group.ID,
+	})
+
+	groupLock := newGroupLock(group.ID, s.instanceID, s.store, logger)
+	if !groupLock.TryLock() {
```

Comment:
Do we want to log an error here? Or a warning?

Reply:
The lock objects have a logger in them already and log on error. Example: https://github.com/mattermost/mattermost-cloud/pull/176/files#diff-2e128f917c08be9b9e3f93249985f459R31

Reply:
For more context, the lockers are usually deferred so we have them handle their own logging needs.

Reply:
Perfect